### ISSUE TITLE
Any possibility of default menus and toolbars in neovim-qt (like in gvim) ?

### ISSUE THREAD

**Issue by aks2161989** [equalsraf/neovim-qt#395]
From a beginner POV, can neovim-qt have default menus and toolbars like that in gvim?

**Comment by equalsraf** [comment]
Menus are possible (neovim-dot-app implements this) I'm not sure if toolbars are possible without changing the core neovim.

**Comment by jeansagi** [comment]
Really?
Is that not possible?
I mean what is the purpose of neovim-qt then?

J.

**Comment by paulondc** [comment]
@jeansagi I don't know what was the initial intention of this project. However, I can say based on my own experience why I started using neovim-qt:

This project provides a consistent neovim experience across multiple platforms by avoiding the issues of running neovim through a console application which by itself can have hotkey conflicts, font rendering issues, among other things that can be time consuming to debug/hard to reproduce.

Although, I appreciate the effort of trying to use qt widgets for some features like tabs, the reasons above are more than enough for me to keep using neovim-qt. I'm not against of menus or toolbars if they can be disabled during startup. But, I find them questionable since, they go in direct conflict with usability in vim by providing features that would require a mouse. IMO, beginner users should learn the proper way of using vim rather than keeping addictive of using a mouse for simple things such as undo, redo, opening&saving a file (@aks2161989).

**Comment by equalsraf** [comment]
To clarify my statement from earlier

- Menus are possible (neovim-dot-app implements this), but they require a good bit of work.
- Based on `:h gui-toolbar` toolbar support should also be possible if menus are already supported, but I have not looked into this any further

> I mean what is the purpose of neovim-qt then?

I usually try to avoid features that overlap with the already provided in nvim core. For example, if you really need a custom toolbar, you could easily take this library and add a toolbar (or other widgets), but adding a completely new system for menu or toolbar setup is completely out of scope since Neovim already provides one.

As a rule of thumb, If I consider that a new feature makes sense I usually try to generalise it and push it into upstream neovim first. We could probably implement a custom toolbar system, but I'd rather do it upstream or wait for solutions to emerge for this purpose.

For the particular case of the menu/toolbar support

- This feature is welcome, if anything I lack the time to implement it myself.
- nvim removed support for some of the toolbar options (`h toolbar` or `toolbariconsize`) those would need to be push upstream if we want them

> Although, I appreciate the effort of trying to use qt widgets for some features like tabs, the reasons above are more than enough for me to keep using neovim-qt.

Thanks to the changes in neovim and https://github.com/equalsraf/neovim-qt/pull/431 this is turning out well. You can now disable/enable tabs in the cli, and once that PR is merged you can also do it from ginit.vim (and some code is already in place for other features).

> I'm not against of menus or toolbars if they can be disabled during startup.

My primary concern is to avoid duplication of work which is already carried out by neovim, Tabs are relatively easy to do. The popupmenu is more complex but it is fully specified upstream.

Menus and other widgets are harder because you need to think carefully how this will interact with neovim e.g. how do you return focus to nvim, or what is expected to happen when you use your mouse or some key accelerator.

**Comment by paulondc** [comment]
@equalsraf The points you raised are super valid. Some of these features they may look super simple, but there is a lot of "unknown" that shows-up when supporting them across multiple platforms, increasing the complexity of the project for features that may not be useful for most users. Your suggestion about using nvim-qt as library is fairly reasonable.

**Comment by Remiol** [comment]
I'm not using neovim till there is a possibility to create my own menu (like menu.vim in gvim). Will there be any possibility in neovim in the future to do that?

**Comment by Remiol** [comment]
@equalsraf will there be a possibility to create menus in neovim in a next release (like menu.vim in gvim)?

**Comment by equalsraf** [comment]
@Remiol I am not working on it. PRs are welcome though.

**Comment by Remiol** [comment]
> 
> 
> @jeansagi I don't know what was the initial intention of this project. However, I can say based on my own experience why I started using neovim-qt:
> 
> This project provides a consistent neovim experience across multiple platforms by avoiding the issues of running neovim through a console application which by itself can have hotkey conflicts, font rendering issues, among other things that can be time consuming to debug/hard to reproduce.
> 
> Although, I appreciate the effort of trying to use qt widgets for some features like tabs. However, the reasons above are more than enough for me to keep using neovim-qt. I'm not against of menus or toolbars if they can be disabled during startup. But, I find them questionable since, they go in direct conflict with the usability in vim by providing features that would require a mouse. IMO, beginner users should learn the proper way of using vim rather than keeping addictive of using a mouse for simple things such as undo, redo, opening&saving a file (@aks2161989).

Undo, redo, opening&saving a file???
There are also users who added their own menu options.

**Comment by nash900** [comment]
This plugin enables you to make your own menu: https://github.com/skywind3000/vim-quickui

**Comment by Remiol** [comment]
> This plugin enables you to make your own menu: https://github.com/skywind3000/vim-quickui

We already can create our own menu in vim using menu.vim
We can't create our menu in neovim (using menu.vim).

**Comment by lmna** [comment]
Gvim toolbar is very handy when using non-latin keyboard layout: in gvim you don`t have to switch to latin to save a file.
